libclc/utils/libclc-remangler/LibclcRemangler.cpp

```diff
@@ -147,340 +123,606 @@ class BumpPointerAllocator {
-      name = StringView(it->second);
-
-    if (name == "void")
-      OB << 'v';
-    else if (name == "wchar_t")
-      OB << 'w';
-    else if (name == "bool")
-      OB << 'b';
-    else if (name == "char")
-      OB << 'c';
-    else if (name == "signed char")
-      OB << 'a';
-    else if (name == "unsigned char")
-      OB << 'h';
-    else if (name == "short")
-      OB << 's';
-    else if (name == "unsigned short")
-      OB << 't';
-    else if (name == "int")
-      OB << 'i';
-    else if (name == "unsigned int")
-      OB << 'j';
-    else if (name == "long")
-      OB << 'l';
-    else if (name == "unsigned long")
-      OB << 'm';
-    else if (name == "long long")
-      OB << 'x';
-    else if (name == "unsigned long long")
-      OB << 'y';
-    else if (name == "__int128")
-      OB << 'n';
-    else if (name == "unsigned __int128")
-      OB << 'o';
-    else if (name == "float")
-      OB << 'f';
-    else if (name == "double")
-      OB << 'd';
-    else if (name == "long double")
-      OB << 'e';
-    else if (name == "__float128")
-      OB << 'g';
-    else if (name == "...")
-      OB << 'z';
-    // TODO: u
-    else if (name == "decimal64")
-      OB << "Dd";
-    else if (name == "decimal128")
-      OB << "De";
-    else if (name == "decimal32")
-      OB << "Df";
-    else if (name == "decimal16")
-      OB << "Dh";
-    else if (name == "char32_t")
-      OB << "Di";
-    else if (name == "char16_t")
-      OB << "Ds";
-    else if (name == "char8_t")
-      OB << "Du";
-    else if (name == "_Float16")
-      OB << "DF16_";
-    else if (name == "auto")
-      OB << 'a';
-    else if (name == "decltype(auto)")
-      OB << 'c';
-    else if (name == "std::nullptr_t")
-      OB << 'n';
-    // Enum
-    else
-      remangleOpenCLCName(typeName, OB, /* Substitutable= */ true);
-  }
+    // If not templated, finish here.
+    if (TemplateArgTys.empty()) {
+      clang::QualType const FuncType =
+          ContextAST->getFunctionType(RetTy, ArgTys, Info);
+      FD->setType(FuncType);
+      FD->setParams(ArgParams);
+      return FD;
+    }
 
-  void remangleOpenCLCQualifiers(const itanium_demangle::Qualifiers quals,
-                                 OutputBuffer &OB) {
-    if (quals & QualConst)
-      OB << "K";
-    if (quals & QualVolatile)
-      OB << "V";
-    if (quals & QualRestrict)
-      OB << "r";
+    // Use FD as a base for a future function specialisation.
+    FunctionDecl *FDSpecialization = FunctionDecl::Create(
+        *ContextAST, ContextAST->getTranslationUnitDecl(), SourceLocation{},
+        DeclarationNameInfo(), VoidFuncType,
+        ContextAST->getTrivialTypeSourceInfo(ContextAST->VoidTy), SC_None,
+        false, false, false, ConstexprSpecKind::Unspecified, nullptr);
+    FDSpecialization->setImplicitlyInline(false);
+
+    FDSpecialization->setDeclName(&ContextAST->Idents.get(KernelName));
+
+    // Used to populate all the template type param decls into.

```

Comment:
Why do we put the type parm decls into a dummy namespace? They should be scoped to the function template right?

Reply:
You're right, that's an early development residue, fixed now.